sandboxjs/fetch_sandbox.js

```diff
@@ -0,0 +1,115 @@
+const branchInfo = require('./bamboo_plan_info.json');
+const path = require('path');
+var xml2js = require('xml2js');
+const fs = require('fs-extra');
+const fetch = require('node-fetch');
+const unzip = require('./unzip');
+
+// Get CDAP version from the pom file.
+async function getCDAPVersion() {
+  return new Promise((resolve, reject) => {
+    var parser = new xml2js.Parser();
+    fs.readFile(path.join(__dirname , '..', 'pom.xml'), function(err, data) {
+      if (err) {
+        reject(err);
+      }
+      parser.parseString(data, function (err, result) {
+        if (err) {
+          reject(err);
+        } 
+        const version = result.project.parent[0].version;
+        resolve(version);
+      });
+    });
+  });
+}
+
+async function downloadSDK(res, pathToZipFile) {
+  const fileStream = fs.createWriteStream(pathToZipFile);
+  return new Promise((resolve, reject) => {
+    res.body.pipe(fileStream);
+    res.body.on("error", reject);
+    fileStream.on("finish", () => {
+      console.log('Done downloading!');
+      resolve();
+    });
+  });
+}
+
+async function fetchSandbox(targetDir) {
+  // Get the last successful build from appropriate plan branch. This plan branch corresponds to bamboo plan name for respective branches
+  const branchResultsRaw = await fetch(`https://builds.cask.co/rest/api/latest/result/${branchInfo.planName}.json?buildstate=successful&max-result=1`);
+  const branchResults = await branchResultsRaw.json();
+  if (!branchResults.results.result.length) {
+    throw `No successful build found in ${branchInfo.planName}`;
+  }
+  const latestSuccessfulBuildName = branchResults.results.result[0].key;
+  let cdapversion;
+  try {
+    cdapversion = await getCDAPVersion();
+    console.log('CDAP version: ', cdapversion);
+  } catch(e) {
+    throw 'Unable to fetch CDAP version from pom.xml file';
+  }
+
+  // Based on the version of CDAP construct the path to the SDK zip file from bamboo build plan.
+  const SDKZipPath = `https://builds.cask.co/browse/${latestSuccessfulBuildName}/artifact/shared/SDK/cdap/cdap-standalone/target/cdap-sandbox-${cdapversion}.zip`;
+  console.log('SDK zip path: ', SDKZipPath);
+  let res;
+  try {
+    res = await fetch(SDKZipPath);
+  } catch(e) {
+    throw `Invalid url to SDK zip file: ${SDKZipPath} - ${e}` 
+  }
+  console.log('fetch begun');
+
+  // Download and unzip SDK
+  const downloadPath = await fs.mkdtemp('cdf_sandbox_download_');
+  const pathToZipFile = path.join(downloadPath, `cdap-sandbox-${cdapversion}.zip`);
+  console.log(`Downloading sandbox to: ${pathToZipFile}`);
+  let unzippedPath;
+  try {
+    await downloadSDK(res, pathToZipFile);
+    console.log('Download complete; unzipping');
+    unzippedPath = await fs.mkdtemp('cdf_sandbox_unzipped_');
+    await unzip(pathToZipFile, unzippedPath);
+    // Move the unzipped files so subsequent scripts don't need to know the version
+    await fs.move(path.join(unzippedPath, `cdap-sandbox-${cdapversion}`), targetDir);
```

Comment:
was learning from Vitalli, that perhaps instead of moving files, we could do a symlink for optimization. Doesn't have to happen here, but is that possible?

Reply:
I tried this but couldn't get it to work. I'll add a note to our optimization ticket. The step with this command only takes 29 seconds though.